contracts/test/PlonkVerifier.t.sol

```diff
@@ -0,0 +1,432 @@
+            uint256 eval,
+            BN254.G1Point memory scalarsAndBasesProd,
+            BN254.G1Point memory openingProof,
+            BN254.G1Point memory shiftedOpeningProof
+        ) = abi.decode(
+            result,
+            (uint256, uint256, uint256, uint256, BN254.G1Point, BN254.G1Point, BN254.G1Point)
+        );
+
+        assertEq(info.u, u);
+        assertEq(info.evalPoint, evalPoint);
+        assertEq(info.nextEvalPoint, nextEvalPoint);
+        assertEq(info.eval, eval);
+        // NOTE: since we cannot directly compare `struct ScalarsAndBases`, we compare their MSM
+        assertEqG1Point(BN254.multiScalarMul(info.commBases, info.commScalars), scalarsAndBasesProd);
+        assertEqG1Point(info.openingProof, openingProof);
+        assertEqG1Point(info.shiftedOpeningProof, shiftedOpeningProof);
+    }
+}
+
+contract PlonkVerifier_computeChallenges_Test is PlonkVerifierCommonTest {
+    /// forge-config: default.fuzz.runs = 5
+    /// @dev Test `computeChallenges` matches that of Jellyfish
+    function testFuzz_computeChallenges_matches(
+        uint64 seed,
+        uint256[] memory publicInput,
+        bytes memory extraTranscriptInitMsg
+    ) external {
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(seed);
+        publicInput = sanitizeScalarFields(publicInput);
+
+        string[] memory cmds = new string[](6);
+        cmds[0] = "diff-test";
+        cmds[1] = "plonk-compute-chal";
+        cmds[2] = vm.toString(abi.encode(vk));
+        cmds[3] = vm.toString(abi.encode(publicInput));
+        cmds[4] = vm.toString(abi.encode(proof));
+        cmds[5] = vm.toString(abi.encode(extraTranscriptInitMsg));
+
+        bytes memory result = vm.ffi(cmds);
+        (V.Challenges memory chal) = abi.decode(result, (V.Challenges));
+
+        V.Challenges memory c = V._computeChallenges(vk, publicInput, proof, extraTranscriptInitMsg);
+        assertEq(chal.alpha, c.alpha);
+        assertEq(chal.alpha2, c.alpha2);
+        assertEq(chal.alpha3, c.alpha3);
+        assertEq(chal.beta, c.beta);
+        assertEq(chal.gamma, c.gamma);
+        assertEq(chal.zeta, c.zeta);
+        assertEq(chal.v, c.v);
+        assertEq(chal.u, c.u);
+    }
+}
+
+contract PlonkVerifier_prepareEvaluations_Test is PlonkVerifierCommonTest {
+    /// forge-config: default.fuzz.runs = 5
+    /// @dev Test if combinng the polynomial evaluations into a single evaluation is done correctly
+    /// is done correctly
+    function testFuzz_prepareEvaluations_matches(
+        uint64 seed,
+        uint256 linPolyConstant,
+        uint256[30] memory scalars
+    ) external {
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(seed);
+        linPolyConstant = sanitizeScalarField(linPolyConstant);
+        uint256[] memory commScalars = sanitizeScalarFields(copyCommScalars(scalars));
+
+        string[] memory cmds = new string[](5);
```

Comment:
Would it be worth writing a function for preparing this command line and calling `vm.ffi` in order to reduce code duplication?

Reply:
Not really imo. Few reasons:
- varying number of arguments: some commands offer more (this might be fine with Solidity's polymorphism support)
- varying argument types: some arguments are `bytes32` (if they are abi encoded), some are `uint256`, some are `uint64`
- varying return types: while the result of `vm.ffi() -> bytes memory` is always `bytes`, when you need concrete/test-specific info for abi decoder
- finally, I find it clearer to write this way and not add too many re-direction. (Optimism's code also does it explicitly without any helper function)